library/pk.c

```diff
@@ -311,24 +316,31 @@ int mbedtls_pk_encrypt( mbedtls_pk_context *ctx,
 int mbedtls_pk_check_pair( const mbedtls_pk_context *pub, const mbedtls_pk_context *prv )
 {
     if( pub == NULL || pub->pk_info == NULL ||
-        prv == NULL || prv->pk_info == NULL ||
-        prv->pk_info->check_pair_func == NULL )
+        prv == NULL || prv->pk_info == NULL )
     {
         return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
     }
 
+    if( pub->pk_info == prv->pk_info && pub->pk_ctx == prv->pk_ctx )
+        return( 0 );
+
+    if( prv->pk_info->check_pair_func == NULL )
+    {
+        return( MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE );
+    }
+
     if( prv->pk_info->type == MBEDTLS_PK_RSA_ALT )
     {
         if( pub->pk_info->type != MBEDTLS_PK_RSA )
             return( MBEDTLS_ERR_PK_TYPE_MISMATCH );
     }
-    else
+    else if( prv->pk_info->type != MBEDTLS_PK_OPAQUE )
     {
         if( pub->pk_info != prv->pk_info )
```

Comment:
Do we still need this check now that `check_pair_func` has access to the entire `pk_context`, and in particular to the `pk_type` of the public key?

Reply:
Why would we remove this check? For transparent objects, must definitely ensure that the base types are the same before calling the function, and changing the check from “same base type” to “same type” would be a behavior change that I see no use for.

Reply:
I don't think it's sensible to enforce a particular PK type in the wrapper function while at the same time widening the scope of `check_pair` by allowing the public key context to be any abstract `pk_context` instance. With the full public key context provided to `check_pair`, I'd prefer the individual implementations to check if they support different types than their own (e.g., `PK_RSA_ALT` supports precisely `PK_RSA`).

Reply:
Good point with `RSA_ALT`. At this point it's the only example, but there might be an `ECKEY_ALT` in the future. Opaque engines need to do the check on their own anyway, and what they're supposed to do with an opaque public key is still an open question.

Reply:
I'm not sure I understand why we would want an `ECKEY_ALT` in the future, as I'm under the impression `OPAQUE` from this PR is more generic. What am I missing?

Reply:
Opaque does dynamic dispatch. It's expected to cohabit with the “native” implementation, and intended for off-chip computations. Alt implementations change the native implementation from software to hardware-accelerated. A device which both some crypto acceleration, and talks to another cryptographic module (e.g. secure element), could have both ALT implementations for its own acceleration, and an opaque engine talking to the secure element.

Mbed TLS doesn't support opaque engines for public key operations and probably won't for a long time. You can call `mbedtls_pk_verify` on an opaque key, but there's no support for that in the X.509 and TLS modules: they don't try to create an opaque key object for a public key obtained from X.509 parsing or via TLS, and they expect to be able to extract all the data from a public key. So opaque keys aren't a viable way to accelerate public key operations.

The reason we might not get `ECKEY_ALT` is that it happens at a lower level with `ECP_ALT`.

Reply:
I'm not sure we're talking about the same thing here: I meant `MBEDTLS_PK_RSA_ALT` and an hypothetic future `MBEDTLS_PK_ECKEY_ALT` in `mbedtls_pk_type_t`, not `MBEDTLS_RSA_ALT` in `config.h` and the bunch of `MBEDTLS_ECXXX_ALT` that are being added - the former does dynamic dispatch and only handles private operations just like Opaque, the later indeed serves a different purpose than Opaque and is quite orthogonal to the PK module in general.

It seems to me that the existing `MBEDTLS_PK_RSA_ALT` should be deprecated in favour of Opaque, so adding a new `MBEDTLS_PK_ECKEY_ALT` along the same lines would seem weird to me.

Reply:
Should I remove the opaque check then?

Reply:
If we deprecate `PK_RSA_ALT` then I think we should keep a type check. But it should be the representation type, not the `type` field, for the sake of `ECKEY`/`ECDH`/`ECDSA`.

Reply:
@gilles-peskine-arm I still feel that this check violates the added flexibility of having a generic PK context in the `check_pair` function, and I don't see the reason why we don't want to make the (representation-)type checking PK-context-specific.

Reply:
I've moved the RSA_ALT check to the rsa_alt_check_pair function, but I'm not sure what else should I change - the non-opaque check is relevant for all non-opaque keys, so should I duplicate it in both rsa_check_pair_wrap and eckey_check_pair? Please review my changes regarding this.

Reply:
@mpg What's your take on this?

Reply:
@hanno-arm  I'm not sure I really understand what you're suggesting and what benefit you think it would bring. Can you please elaborate?

Reply:
[After chatting on another medium.] @hanno-arm @gilles-peskine-arm I think there are actually to orthogonal issues here: (1) whether check for type should be done here or in individual `xxx_check_pair()` functions (in `pk_wrap.c`), and (2) whether we should check for strict type equality (except for `RSA_ALT`) or for equality of representation types.

Regarding (1), I think it would be cleaner in principle to have the type check in `pk_wrap.c`, as the current philosophy of the module is to avoid type-specific checks (such as the one for `RSA_ALT` here) in `pk.c` and have them in `pk_wrap.c` instead. On the other hand, this might mean some code duplication as all `xxx_check_pair()` functions (except `rsa_alt_`) are going to start with similar type checks.

Regarding (2), representation type would probably make more sense, but it's also a change of behaviour (even though existing behaviour wasn't documented to that level of detail) and nobody's complaining about the current behaviour.

I actually don't have a strong opinion regarding either of those questions, I think both ways are OK for each question.

Reply:
`pk.c` doesn't make assumptions about type-specific things, but `pk.h` does, via the peculiar choice of cases in `mbedlts_pk_type_t`.

On balance, I'm inclined towards keeping the existing behavior. In `pk.c`, in the dispatcher for `mbedtls_pk_check_pair`, enforce that either the types must be the same or the private key must be an opaque type (`RSA_ALT` or `OPAQUE`) and then the public key may be of a different type. It may be less clean, but it's more robust.

@hanno-arm The reason I don't like distributing the type check over the individual `check_pair_func` implementations is that it's less robust: you'd then need to get it right in all of them, not just in one place.

Reply:
@gilles-peskine-arm Ok, I feel we have exchanged the arguments, and it's not as important as to block the PR because of this. We can go on with your preference.

Reply:
Updated to the behaviour described by Gilles, with an update of the API.